# **NumPy Linear Algebra**


**NumPy** is a powerful library in Python used for scientific computing, especially for working with arrays and matrices. One of the key features of NumPy is its support for linear algebra operations.

Linear algebra is a branch of mathematics that deals with linear equations, matrices, and vectors. In NumPy, linear algebra operations are performed using the linalg module. This module contains a wide range of functions for performing various linear algebra operations, such as matrix multiplication, solving linear equations, finding determinants and eigenvalues, and performing singular value decomposition (SVD).

Using NumPy's linear algebra capabilities, you can efficiently and accurately perform complex calculations on matrices and arrays, making it a valuable tool for scientific computing and data analysis.

NumPy provides the following functions to perform the different algebraic calculations on the input data:

| SN | Function  | Definition |
|:---------|:--------:|---------:|
|   1   |  dot()   |  It is used to calculate the dot product of two arrays.   |
|   2   |  vdot()  |  It is used to calculate the dot product of two vectors.  |
|   3   |  inner()   |  It is used to calculate the inner product of two arrays.   |
|   4   |  matmul()  | It is used to calculate the matrix multiplication of two arrays.   |
|   5   |  det()   | It is used to calculate the determinant of a matrix.   |
|   6   |  solve()  |  It is used to solve the linear matrix equation.  |
|   7   |  inv()   |  It is used to calculate the multiplicative inverse of the matrix.   |

# 1. numpy.dot() function
    
The numpy.dot() function is a part of NumPy's linear algebra module that is used to perform matrix multiplication between two arrays. It takes two arrays as input and returns their dot product.

The dot product of two arrays is calculated by multiplying the corresponding elements of the arrays and then summing them up. The dot product of two arrays 
𝐴 and 
𝐵
 can be denoted by 
𝐴⋅𝐵 or  𝐴∗𝐵.

**The numpy.dot()** function can be used with arrays of different shapes, as long as the dimensions of the arrays are compatible for matrix multiplication. Specifically, the number of columns in the first array must be equal to the number of rows in the second array. If the shapes of the arrays are not compatible for matrix multiplication, the function will raise a ValueError.

**Example 01:**

In [33]:
import numpy as np
a = np.array([100, 200])
b = np.array([23, 12]) 
print ("original Array")
print(a)
print(b)
dot = np.dot(a, b)
print("Dot product: ",dot)


original Array
[100 200]
[23 12]
Dot product:  4700


**Explanation:**

In this example, two 1D arrays a and b are defined.
    
The numpy.dot() function computes the dot product, resulting in a scalar value.

    Dot product: 100 * 23 + 200 * 12 = 4600

**Example 02**

In [13]:
import numpy as np
a = np.array([[100, 200], [23, 12]]) 
b = np.array([[10, 20], [12, 21]])  

dot = np.dot(a, b)  
print("Values of dot product:", dot)  


Values of dot product: [[3400 6200]
 [ 374  712]]


Explanation:

This code defines two 2D NumPy arrays a and b, each with 2 rows and 2 columns.
The numpy.dot() function performs matrix multiplication between a and b.
The values of the resulting array dot are calculated as follows:

dot[0,0] = (100 * 10) + (200 * 12) = 3400

dot[0,1] = (100 * 20) + (200 * 21) = 6200

dot[1,0] = (23 * 10) + (12 * 12) = 374

dot[1,1] = (23 * 20) + (12 * 21) = 712

# 2. numpy.vdot() function
    
**The numpy.vdot()** function computes the dot product of two 1D arrays. It returns the dot product of the two input arrays, which is computed as the sum of the element-wise product of the two arrays. This function is specifically designed to work with 1D arrays.


Example 01:

In [37]:
import numpy as np  
a = np.array([[100, 200], [23, 12]])  
b = np.array([[10, 20], [12, 21]]) 

vdot = np.vdot(a, b)  
print("Values of vector dot product:", vdot)


Values of vector dot product: 5528


**Explanation:**

Even though a and b are 2D arrays, numpy.vdot() treats them as flattened vectors.

Flattened a = [100, 200, 23, 12], b = [10, 20, 12, 21]

Vector dot product: 100 * 10 + 200 * 20 + 23 * 12 + 12 * 21 = 5528

In [43]:
import numpy as np

a = np.array([1 + 2j, 3 + 4j])
b = np.array([5 + 6j, 7 + 8j])

vdot = np.vdot(a, b)
print(vdot)


(70-8j)


**Explanation:**

The function computes the vector dot product of complex numbers.

Conjugate of a: [1 - 2j, 3 - 4j]

Vector dot product: (1 - 2j) * (5 + 6j) + (3 - 4j) * (7 + 8j) = 70 - 8j

# 3. numpy.inner() function:

The **numpy.inner()** function calculates the inner product of two arrays. For 1D arrays, it computes the dot product. For higher-dimensional arrays, it sums the product over the last dimension of the first array and the second-to-last dimension of the second array.


Example 01:

In [47]:
import numpy as np
a = np.array([1, 2, 3, 4, 5, 6])  
b = np.array([23, 23, 12, 2, 1, 2])  

inner = np.inner(a, b)  
print(inner) 


130


**Explanation:**

The numpy.inner() function sums the product of corresponding elements.

    Inner product: 1*23 + 2*23 + 3*12 + 4*2 + 5*1 + 6*2 = 130

**Example 02**

In [54]:
import numpy as np
a = np.array([[1, 2, 3], [4, 5, 6]])  
b = np.array([[23, 23, 12], [2, 1, 2]])  

inner = np.inner(a, b)  
print("Values of inner product:", inner)


Values of inner product: [[105  10]
 [279  25]]


**Explanation:**

The inner product is calculated for each row of a and each row of b.
    
Resulting array inner:

inner[0,0] = (1*23) + (2*23) + (3*12) = 105

inner[0,1] = (1*2) + (2*1) + (3*2) = 10

inner[1,0] = (4*23) + (5*23) + (6*12) = 279

inner[1,1] = (4*2) + (5*1) + (6*2) = 25

# 4. numpy.matmul() function:

The **numpy.matmul()** function performs matrix multiplication. It multiplies two arrays using matrix multiplication rules, where the number of columns in the first array must match the number of rows in the second array.


    
Example 01: Matrix product

In [61]:
import numpy as np
a = np.array([[100, 200], [23, 12]]) 
b = np.array([[10, 20], [12, 21]])  

mul = np.matmul(a, b)  
print("Matrix multiplication:", mul) 


Matrix multiplication: [[3400 6200]
 [ 374  712]]


**Explanation:**

Matrix multiplication using numpy.matmul().

Calculated similarly as with numpy.dot().

**Example 02: Element-wise matrix multiplication**

In [70]:
import numpy as np  
a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])  
b = np.array([[9, 8, 7], [6, 5, 4], [3, 2, 1]]) 

mul = np.multiply(a, b)  
print("Element-wise multiplication:", mul)  


Element-wise multiplication: [[ 9 16 21]
 [24 25 24]
 [21 16  9]]


**Explanation:**

Element-wise multiplication: Each element of a is multiplied by the corresponding element of b.

Resulting array mul:

mul[0,0] = 1*9, mul[0,1] = 2*8, etc.

# 5. numpy.linalg.det() function
    
The numpy.linalg.det() function calculates the determinant of a matrix. The determinant is a scalar value that is a function of the entries of a square matrix. It can be used to determine if a matrix has an inverse and is used in various matrix-related calculations.


    
**Example 01:**

In [75]:
import numpy as np  
a = np.array([[3, 6], [1, 3]]) 

det = np.linalg.det(a)  
print(det) 


3.0000000000000004


**Explanation:**

This code calculates the determinant of a 2x2 matrix a.

Determinant formula: det(a) = (3*3) - (6*1) = 3

In [79]:
import numpy as np  
a = np.array([[100, 20], [10, 15]]) 

det = np.linalg.det(a)  
print(det)  


1300.0


# 6. numpy.linalg.solve() function
    
The **numpy.linalg.solve()** function solves a system of linear equations of the form Ax = B, where A is a square matrix and B is a matrix with the same number of rows as A. The function returns the solution x.




**Example 01:**

In [83]:
import numpy as np  
a = np.array([[5, 3], [3, 2]])  
b = np.array([[5], [2]])  

sol = np.linalg.solve(a, b)  
print(sol)  


[[ 4.]
 [-5.]]


**Explanation:**

This code solves the system of equations:

5x+3y=5
and
3x+2y=2

Result: x = 4, y = -5

# 7. numpy.linalg.inv() function
    
The **numpy.linalg.inv()** function calculates the multiplicative inverse of a matrix. If A is a square matrix, then the inverse of A is a matrix B such that AB = BA = I, where I is the identity matrix.



**Example:**


In [97]:
import numpy as np  
a = np.array([[1, 2], [3, 4]])  

inv = np.linalg.inv(a)  
print(inv)  


[[-2.   1. ]
 [ 1.5 -0.5]]


**Explanation:**

This code calculates the inverse of the 2x2 matrix a.

Resulting matrix inv:

Inverse of a: [[ -2, 1], [ 1.5, -0.5 ]]

These examples and explanations provide an overview of various linear algebra functions available in NumPy, allowing you to perform a wide range of mathematical operations efficiently and effectively.
